bee-common/src/logger/mod.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright 2020 IOTA Stiftung
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and limitations under the License.
+
+mod config;
+
+pub use config::{LoggerConfig, LoggerConfigBuilder, LoggerOutputConfig, LoggerOutputConfigBuilder};
+
+use fern::{
+    colors::{Color, ColoredLevelConfig},
+    Dispatch,
+};
+
+/// Name of the standard output.
+pub const LOGGER_STDOUT_NAME: &str = "stdout";
+
+/// Error occuring when initializing a logger backend.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum Error {
+    CreatingFileFailed,
+    InitializationFailed,
+}
+
+/// Initializes a `fern` logger backend for the `log` crate.
+///
+/// # Arguments
+///
+/// * `config`  -   Logger configuration
+pub fn logger_init(config: LoggerConfig) -> Result<(), Error> {
+    let timestamp_format = "[%Y-%m-%d][%H:%M:%S]";
+
+    let mut logger = if config.color {
+        let colors = ColoredLevelConfig::new()
+            .trace(Color::BrightMagenta)
+            .debug(Color::BrightBlue)
+            .info(Color::BrightGreen)
+            .warn(Color::BrightYellow)
+            .error(Color::BrightRed);
+
+        // Creates a logger dispatch with color support.
+        Dispatch::new().format(move |out, message, record| {
+            out.finish(format_args!(
+                "{}[{}][{}] {}",
+                chrono::Local::now().format(timestamp_format),
+                record.target(),
+                colors.color(record.level()),
+                message
+            ))
+        })
+    } else {
+        // Creates a logger dispatch without color support.
+        Dispatch::new().format(move |out, message, record| {
+            out.finish(format_args!(
+                "{}[{}][{}] {}",
+                chrono::Local::now().format(timestamp_format),
+                record.target(),
+                record.level(),
+                message
+            ))
+        })
```

Comment:
this can be made DRY

Reply:
Only in the sense that the output `Dispatch::new().format` bit can be removed. The inner `format_args!` uses different types for the arguments and so it's not possible to whack a `if` in there for the argument that differs.

Reply:
alright :+1:

Reply:
How can I even remove `Dispatch::new().format` ?

Reply:
why the obvious way of creating a macro not work?

Reply:
You can put the `if` inside `out.finish` instead of surrounding the entry expression.

Reply:
Not that it really matters anyway, I don't have a problem with things as they are.

Reply:
`log_dispatch!` maybe?

Reply:
A macro is not even necessary. Why not just create a `create_dispatch(...) -> Dispatch` function?

Reply:
And what would it look like ? The format itself is different.

Reply:
ah, right...., but you can definitely solve this with a macro.

Reply:
then it will just replace the expression `colors.color(record.level())` with `record.level()`.
Or am I being stupid right now? could be

Reply:
okay, you could write a macro `log_format!(timestamp_format, record, message)`, that at least DRYs the concrete format string, so it only needs to be changed once in case we decide to.

Reply:
this macro is called instead of the `format_args!` macro, which now is called inside of `log_format!`

Reply:
sorry Thibault. I think we should not spend too much time on this. We can simplify that later. What you at least could consider doing is put the format string "{}[{}][{}] {}" as a constant, so we only need to change that in one place if we decide to.